src/server/pfs/s3/multipart.go

```diff
@@ -0,0 +1,553 @@
+		if err != nil {
+			logrus.Errorf("could not remove uploadID=%s, partNumber=%d: %v", uploadID, partNumber, err)
+		}
+	}
+}
+
+func (h *multipartHandler) complete(w http.ResponseWriter, r *http.Request) {
+	repo, branch, _ := objectArgs(r)
+	branchInfo, err := h.pc.InspectBranch(repo, branch)
+	if err != nil {
+		newNotFoundError(r, err).write(w)
+		return
+	}
+	uploadID := r.FormValue("uploadId")
+	if err := h.multipartManager.checkExists(uploadID); err != nil {
+		newNoSuchUploadError(r).write(w)
+		return
+	}
+
+	name, err := h.multipartManager.filepath(uploadID)
+	if err != nil {
+		newInternalError(r, err).write(w)
+		return
+	}
+
+	bodyBytes, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		err = fmt.Errorf("could not read request body: %v", err)
+		newInternalError(r, err).write(w)
+		return
+	}
+
+	payload := CompleteMultipartUpload{}
+	err = xml.Unmarshal(bodyBytes, &payload)
+	if err != nil {
+		newMalformedXMLError(r).write(w)
+		return
+	}
+
+	// verify that there's at least part, and all parts are in ascending order
+	isSorted := sort.SliceIsSorted(payload.Parts, func(i, j int) bool {
+		return payload.Parts[i].PartNumber < payload.Parts[j].PartNumber
+	})
+	if len(payload.Parts) == 0 || !isSorted {
+		newInvalidPartOrderError(r).write(w)
+		return
+	}
+
+	// ensure all the files exist
+	for _, part := range payload.Parts {
+		if err = h.multipartManager.checkChunkExists(uploadID, part.PartNumber); err != nil {
+			newInvalidPartError(r).write(w)
+			return
+		}
+	}
+
+	// pull out the list of part numbers
+	partNumbers := []int{}
+	for _, part := range payload.Parts {
+		partNumbers = append(partNumbers, part.PartNumber)
+	}
+
+	// A reader that reads each file chunk. Because this acquires a lock,
+	// `Close` MUST be called, or the multipart manager will deadlock
+	reader := newMultipartReader(h.multipartManager, uploadID, partNumbers)
+
+	_, err = h.pc.PutFileOverwrite(branchInfo.Branch.Repo.Name, branchInfo.Branch.Name, name, reader, 0)
+	if err != nil {
+		if closeErr := reader.Close(); closeErr != nil {
+			logrus.Errorf("could not close reader for uploadID=%s: %v", uploadID, closeErr)
+		}
+		newInternalError(r, err).write(w)
+		return
+	}
+
+	if err = reader.Close(); err != nil {
+		newInternalError(r, err).write(w)
+		return
+	}
+
+	if err = h.multipartManager.remove(uploadID); err != nil {
+		newInternalError(r, err).write(w)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func (h *multipartHandler) del(w http.ResponseWriter, r *http.Request) {
+	repo, branch, _ := objectArgs(r)
+	_, err := h.pc.InspectBranch(repo, branch)
+	if err != nil {
+		newNotFoundError(r, err).write(w)
+		return
+	}
+	uploadID := r.FormValue("uploadId")
+	if err := h.multipartManager.checkExists(uploadID); err != nil {
+		newNoSuchUploadError(r).write(w)
+		return
+	}
+	if err := h.multipartManager.remove(uploadID); err != nil {
+		newInternalError(r, err).write(w)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+}
```

Comment:
A few thoughts on multipart uploads:
- using the local filesystem has the problem that it will fail when there are multiple pachd nodes because requests will likely get routed to different nodes. @brycemcanally is working on using the local filesystem for normal puts though so we can probably find a common solution to this.
- It might make sense to stream this data directly into object storage rather than using the local filesystem. Given that most of the object stores have a notion of multipart uploads themselves.

Reply:
By streaming the data directly into object storage, do you mean PFS? If so, I'm not sure how we'd do that given the S3 API:

* You upload parts of the file in individual requests, but they can come in out of order or in parallel - e.g. I could upload part 1, then part 3, then part 2.
* The parts are combined/merged when you complete a multipart upload, but you can skip certain parts - e.g. I could combine just part 1 and 3.

Reply:
So I was thinking the underlying object storage, not pfs. Locally could cause some problems, I think we'd need to implement a multipart upload call in our obj api, that way we could avoid needing a roundtrip to the cloud (and needing to get local to have access to a cloud object store which is probably more annoying, and defeats the purpose of local mode.) Anyways, not sure if this is a great idea, just a thought I had.

Reply:
What about just having a separate k8s persistent volume? That way we wouldn't need to implement anything new (well, above and beyond what's already here.)

Reply:
A persistent volume would really only help if you want to be able to pick up where you left off in the event of a pachd crash. That's probably not so important to have, it doesn't solve the problems with multiple pachd servers afaik.

Reply:
Just to clarify, none of this has to be done before this gets merged in. I'm just speculating on how we can get this working with fewer restrictions in the future.

Reply:
You can't mount the same persistent volume on more than one pod?

Reply:
^ TIL that's generally the case. 

What about a separate PFS repo for multipart content?

Reply:
Yeah, you can't have a volume that's shared between multiple pods. I think doing it as a separate repo could work, and get a similar result to writing it directly to the object store, but I think you'll face a bit of an impedance mismatch. You'll either need to write a document per chunk into etcd so they can be represented as files. Or you'll need to do a one-off commit and have a document per commit in etcd. Neither one is super pretty. I think instead you can basically just proxy the multipart upload stuff directly to the object store and then have a single etcd document that keeps track of the existence of multiple part uploads. You might not even need that since multipart uploads already have an id associated with them, you might be able to just proxy directly.